Reboot - add parameter for paths to search for shutdown command

samdoran opened the pull request:
##### SUMMARY

When [`molly-guard`](https://en.wiktionary.org/wiki/molly-guard) is installed, the normal `shutdown` command is moved to `/lib/molly-guard/`. Add a parameter to allow the default search paths to be overridden.

Fixes #51190 

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request


##### COMPONENT NAME
`reboot.py`

samdoran committed:
Look in /lib/molly-guard for shutdown command

Signed-off-by: Sam Doran <sdoran@redhat.com>

samdoran committed:
Add parameter for extra_search_paths

Signed-off-by: Sam Doran <sdoran@redhat.com>

samdoran committed:
Change option to search_paths

- Update docs
- Make the parameter replace the default values
- Add some sanity checking on the data type

Signed-off-by: Sam Doran <sdoran@redhat.com>

nitzmahone commented on test/integration/targets/reboot/tasks/main.yml:
```diff
@@ -37,11 +37,13 @@
     - name: Reboot with all options
       reboot:
         connect_timeout: 30
+        extra_search_paths: /usr/local/bin
```
s/extra_// - and BTW, shouldn't this be failing tests -both because reboot isn't in /usr/local/bin and because `extra_search_paths` isn't a valid arg?

samdoran replied:
Funny thing about these tests is they don't run due to the lack of separation of controller/target in tests. But I keep updating them and running them locally for when we _can_ use them. Evidently I failed to run them before pushing, though. :)

nitzmahone commented on test/integration/targets/reboot/tasks/main.yml:
```diff
@@ -75,6 +78,27 @@
+
+        - import_tasks: check_reboot.yml
+
+      when: ansible_facts.pkg_mgr == 'apt'
```
maybe ansible_distribution in ['debian','ubuntu'] - we don't know if other distros might not have that package...

nitzmahone commented on test/integration/targets/reboot/tasks/main.yml:
```diff
@@ -75,6 +78,27 @@
+
+        - name: Reboot when molly-guard is installed
+          reboot:
+            extra_search_paths: /lib/molly-guard
```
s/extra_//

nitzmahone reviewed (changes requested):
need to fix some of the arg rename stuff in tests, and it appears the tests aren't running or something, because they should be failing in current state...

samdoran committed:
Correct parameter in tests, change conditional for molly-guard tasks

Signed-off-by: Sam Doran <sdoran@redhat.com>

nitzmahone commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -130,21 +131,39 @@ def get_distribution(self, task_vars):
+        if not isinstance(search_paths, list):
+            raise AnsibleError("'search_paths' must be a string or list, got {0}: {1}".format(type(search_paths), search_paths))
+
+        # If we did get a list, ensure it's a flast list and does not contain nested data structures
```
s/flast/flat/

nitzmahone commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -130,21 +131,39 @@ def get_distribution(self, task_vars):
+            raise AnsibleError("'search_paths' must be a string or list, got {0}: {1}".format(type(search_paths), search_paths))
+
+        # If we did get a list, ensure it's a flast list and does not contain nested data structures
+        nested_list = any(isinstance(x, (dict, list)) for x in search_paths)
```
should probably be `all(is_string(x ...` instead; if the path isn't a string, it's a fail, right?

samdoran replied:
This check is looking for lists or dicts in the `search_paths` field:

```
search_paths:
  -
    - /some/path
    - /some/other/path
  hey: there
```

I can add another check to each of the search paths are actually strings.

nitzmahone reviewed (changes requested):
couple minor things

samdoran committed:
Address feedback

nitzmahone commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -142,10 +143,15 @@ def get_shutdown_command(self, task_vars, distribution):
 
-        # If we did get a list, ensure it's a flast list and does not contain nested data structures
+        # If we did get a list, ensure it's a flat list and does not contain nested data structures
         nested_list = any(isinstance(x, (dict, list)) for x in search_paths)
```
redundant with the is_string check below now; just meld those two together (so you're keeping the more declarative "any" rather than the explicit loop) and you're good to go

samdoran replied:
Thanks for the info. I simplified it quite a bit.

nitzmahone reviewed (changes requested):
one last thing

samdoran committed:
Simplify field validation

Needed the try except in case a non-iterable type is put in the search_paths field
